### PR TITLE
Add option to track RMM allocations

### PR DESCRIPTION
**Pull request by shwina**
Adds the `rmm_track_allocations` option that enables workers to query the amount of RMM memory allocated at any time via `mr.get_allocated_bytes()`.

This is used in https://github.com/dask/distributed/pull/5740.

### CODE REVIEW

**Review by pentschev — COMMENTED** [review]
That looks really cool, thanks @shwina . Could we add a test as well? We have some in https://github.com/rapidsai/dask-cuda/blob/b60dec90cf56dbc0937dfdce394200b08e3ebeba/dask_cuda/tests/test_local_cuda_cluster.py#L140-L202 that can be used as reference.

**Review by pentschev — COMMENTED** [review]
@shwina it seems that there are some style issues, I suggest using `pre-commit` to automatically resolve them at commit time.

**Review by pentschev — CHANGES_REQUESTED** [review]
@shwina thanks for addressing those. I have a couple of minor requests, shouldn't be too much work, but otherwise looks great!

**Comment by pentschev** [comment]
Errors are unrelated, opened https://github.com/rapidsai/dask-cuda/issues/861 to track.

File: dask_cuda/tests/test_dask_cuda_worker.py [hunk 1]
```diff
@@ -253,3 +253,34 @@ def test_cuda_visible_devices_uuid(loop):  # noqa: F811
...
+                    rmm.mr.get_current_device_resource_type
+                )
+                for v in memory_resource_type.values():
+                    assert v is rmm.mr.TrackingResourceAdaptor
+
+                memory_resource_upstream_type = client.run(
+                    lambda: type(rmm.mr.get_current_device_resource().upstream_mr)
+                )
+                for v in memory_resource_upstream_type.values():
+                    assert v is rmm.mr.PoolMemoryResource
```
[the remark this answers is not in the document]
Comment (shwina) [reply dask_cuda/tests/test_dask_cuda_worker.py:286]: Thanks! I added that.

File: dask_cuda/local_cuda_cluster.py [hunk 2]
```diff
@@ -210,6 +210,7 @@ def __init__(
         rmm_managed_memory=False,
         rmm_async=False,
         rmm_log_directory=None,
+        rmm_track_allocations=False,
```
[the remark this answers is not in the document]
Comment (shwina) [reply dask_cuda/local_cuda_cluster.py:223]: Done!

**Review by pentschev — APPROVED** [review]
LGTM. There's the failing test as reported in https://github.com/rapidsai/dask-cuda/issues/861 which blocks this PR, if this is not critical to go in immediately, let's wait for that to be fixed, otherwise we may need to xfail that test temporarily.

Thanks @shwina !

**Comment by pentschev** [comment]
Actually there are some style issues too @shwina , could you fix them? I would strongly suggest using `pre-commit` to avoid this hassle for every new commit. 😄

File: dask_cuda/utils.py [hunk 3]
```diff
@@ -56,10 +57,12 @@ def __init__(
         self.async_alloc = async_alloc
         self.logging = log_directory is not None
         self.log_directory = log_directory
+        self.rmm_track_allocations = track_allocations
 
     def setup(self, worker=None):
+        import rmm
```
Comment (charlesbluca) [inline dask_cuda/utils.py:63]: Are we safe to always import RMM here? If this function is always run as part of the `RMMSetup` plugin, wouldn't that make RMM an implicit dependency for starting a CUDA cluster?

Comment (pentschev) [reply dask_cuda/utils.py:63]: Great catch @charlesbluca , indeed I failed to realize that. Yes, we should remove it from the "main" context of `setup` and leave it solely within the local contexts that exist today, iff one of those options to enable RMM is set by the user.

File: dask_cuda/utils.py [hunk 4]
```diff
@@ -56,10 +57,12 @@ def __init__(
...
+
         if self.async_alloc:
```
Comment (pentschev) [inline dask_cuda/utils.py:65]:
```suggestion
        if self.async_alloc:
```

File: dask_cuda/utils.py [hunk 5]
```diff
@@ -83,6 +85,9 @@ def setup(self, worker=None):
                     worker, self.logging, self.log_directory
                 ),
             )
+        if self.rmm_track_allocations:
```
Comment (pentschev) [inline dask_cuda/utils.py:88]:
```suggestion
        if self.rmm_track_allocations:
            import rmm

```

**Review by pentschev — CHANGES_REQUESTED** [review]

File: dask_cuda/utils.py (hunk 3 above)
Comment (shwina) [reply dask_cuda/utils.py]: Thanks both! I've accepted @pentschev's suggestions here.

File: dask_cuda/utils.py [hunk 6]
```diff
@@ -56,10 +57,10 @@ def __init__(
         self.async_alloc = async_alloc
         self.logging = log_directory is not None
         self.log_directory = log_directory
+        self.rmm_track_allocations = track_allocations
 
     def setup(self, worker=None):
         if self.async_alloc:
```
Comment (pentschev) [inline dask_cuda/utils.py:63]:
```suggestion
        if self.async_alloc:
            import rmm
```

File: dask_cuda/utils.py [hunk 7]
```diff
@@ -69,7 +70,6 @@ def setup(self, worker=None):
                     )
                 )
         elif self.initial_pool_size is not None or self.managed_memory:
```
Comment (pentschev) [inline dask_cuda/utils.py:72]:
```suggestion
        elif self.initial_pool_size is not None or self.managed_memory:
            import rmm
```

**Review by pentschev — COMMENTED** [review]
Forgot to add back the previously removed RMM imports.

**Review by pentschev — APPROVED** [review]

**Comment by pentschev** [comment]
LGTM now, thanks @shwina for the work here, and @charlesbluca for the review and fixes here too!

**Comment by pentschev** [comment]
@gpucibot merge

**Comment by pentschev** [comment]
rerun tests

**Comment by pentschev** [comment]
rerun tests

**Comment by pentschev** [comment]
rerun tests

**Comment by pentschev** [comment]
@gpucibot merge

**Pull request merged**